[V3 commands] Introduction to Red's converters + TimeConverter

laggron42 opened the pull request:
### Type

- [ ] Bugfix
- [ ] Enhancement
- [x] New feature

### Description of the changes

This PR adds a `converters.py` file, following d.py's style. This comes with the `TimeConverter` converter, which converts a string to a `datetime.timedelta` unit. It can be called manually or be used as a function annotation, like other converters.

You can provide many types of strings and this will be converted to a valid class, thanks @mikeshardmind for the regex thing.

This PR will be needed for the future timed mute, that should be made by me if everything's alright.

### Example

- `5m` > `timedelta(seconds=300)`
- `10 minutes` > `timedelta(seconds=600)`
- `1 second` > `timedelta(seconds=1)`
- `5d` > `timedelta(days=5)`

### Documentation

![](https://screenshotscdn.firefoxusercontent.com/images/b26967b8-c820-4704-a30a-698bb16c8d89.png)

laggron42 committed:
Custom time converter

Convert a string to a datetime.timedelta unit.

laggron42 committed:
Add documentation

laggron42 committed:
Match discord.py's style

laggron42 committed:
Fix docs reference

mikeshardmind commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,74 @@
+__all__ = ["TimeConverter"]
+
+
+class TimeConverter(Converter):
```
This should probably be `TimedeltaConverter`

mikeshardmind commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,74 @@
+        ~discord.ext.commands.BadArgument
+            No time was found from the given string.
+        """
+        TIME_RE = re.compile(
```
This is a compiled constant, it belongs outside the class at the top of the file

laggron42 replied:
What do you mean? Put arguments outside of the class and keep what's related to time inside the class?

mikeshardmind replied:
You shouldn't be calling `re.compile` each time a conversion happens. Make it a module level constant, but don't place it in `__all__`'s definition.

mikeshardmind commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,74 @@
+        matches = TIME_RE.match(argument)
+        params = {k: int(v) for k, v in matches.groupdict().items() if None not in (k, v)}
+        time = timedelta(**params)
+        if str(time) == "0:00:00":
```
This disallows a zero second timedelta, which could be allowed by checking against `if not params` earlier in the code instead.

laggron42 replied:
But if the user sets, for example `2 weeks`, Regex will return 0:00:00, while the user is actually expecting something else. For example, with timed mute, if the user input a wrong unit, what will happen is that the user will get muted and immediately unmuted.

I'm trying to follow `discord.py`'s style for converters, and when it can't convert, it raises `BadArgument` (instead of returning None or something like that).

mikeshardmind replied:
Your handling of this places `0 seconds` as invalid, when the intent is to not handle `2 weeks`. Meanwhile, if `params` is an empty dict here, you can raise the bad argument there (having matched nothing expected)

mikeshardmind reviewed (changes requested):
I haven't tested the implementation yet, but I already noticed a few things with it. And next time, careful with those GitHub pings.

Tobotimus commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,74 @@
+import re
+from datetime import timedelta
+from . import Converter, Context, BadArgument
```
I know this package is a bit confusing due to the way it imports all of discord.py's stuff to the top level, then writes our stuff afterwards.

When importing something we've overwritten from within the actual `commands` package, import it from the sibling-module, not from the top level. So in this example, it would be `Context`, you want to do `from .context import Context`.

Also, whenever you're doing a relative import which is only for a type-hint, please use [`typing.TYPE_CHECKING`](https://docs.python.org/3.7/library/typing.html#typing.TYPE_CHECKING) and forward references to avoid any circular imports.

laggron42 replied:
okay for `Context`, but what about `Converter` and `BadArgument`? Should I import `discord.ext.commands` instead?

laggron42 committed:
Use Timeconverter instead of Time

laggron42 committed:
Use automodule instead of autoclass

Commit:
Sinbad's changes requested

- use TYPE_CHECKING instead of importing Context
- put TIME_RE outside of the class
- do not block 0:00:00

Zephyrkul commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,80 @@
+            async def convert_time(ctx: Context, text: str) -> datetime.timedelta:
+                time = await commands.Timedelta.convert(ctx, text)
+                return time
+    """
```
Couple issues with these code blocks.
1. lines 39 and 47: it's TimedeltaConverter, not Timedelta.
2. line 48: You're using the class to call an instance's method. You need to use an instance to call an instance method.

laggron42 committed:
Use TimedeltaConverter in docs
Use an instance to call the method

Redjumpman commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,80 @@
+                return time
+    """
+
+    async def convert(self, ctx: "Context", argument: str) -> timedelta:
```
I'm not sure what ctx is doing here. It's not used for the rest of the code in this function, and the "type" checking is just "Context". Did you mean to check for `commands.Context` ?

Zephyrkul replied:
ctx is required for converters that inherit from d.py's Converter. That said, if there's no need for ctx, perhaps just a normal module function would be better. d.py takes any callable as a converter, after all.

laggron42 replied:
I tried to follow what the `discord.py` docs recommand. Please read this, and if you still want to remove that argument, tell me.
https://discordpy.readthedocs.io/en/rewrite/ext/commands/commands.html#advanced-converters

Zephyrkul replied:
the discord.py docs recommend using basic converters unless you require ctx. Since you don't require ctx, then a basic converter will likely be a better approach.
https://discordpy.readthedocs.io/en/rewrite/ext/commands/commands.html#basic-converters

laggron42 replied:
I find it easier for the users to use the advanced converters, because the core discord.py's converters are advanced. The users are used to call a function like that:
```py
await commands.MemberConverter.convert(ctx, text)
```
Using the basic converters would give this:
```py
commands.timedelta_converter(text)
```
Using a different style is quite a mess, especially since **`redbot.core.commands` inherits from `discord.ext.commands`** and include their converters. We should just keep the same style as the other converters. Just imagine:
- For a member, I call `commands.MemberConverter.convert()`
- For a channel, I call `commands.TextChannelConverter.convert()`
- For a role, I call `commands.RoleConverter.convert()`
- For time, I call `commands.timedelta_converter()`

That doesn't match at all with the other converters and will confuse the users. That's why I'm using advanced converters, so it works the same as the other ones, which are often used.

Basic converters may be useful for a personnal cog, but not for an API that already uses an other type of converters under the same module!

Zephyrkul replied:
discord.py includes basic converters as well, ex `clean_content`. More than that, basic converters are all over the bot, just look at any `int` or `str` typehints. Those are basic converters in d.py's eyes as well. 

The advanced converters are advanced *because* they require `ctx`. The advanced converters require `ctx` because they are getting data about the context to return Discord objects. Your converter does not require `ctx` because it's neither returning any Discord objects nor getting any data about Discord's context.

Creating a convoluted method to convert to match Discord converters which *require* that convoluted method to function is quite simply absurd. You should match the functionality of what you require, keep it as simple as possible, and above all actually match how d.py functions with its typehint conversions rather than make nonsense excuses for what you coded.

Besides which, what happens if I want to convert an argument to a timedelta in a Discord event? Your method *requires* me to have a Context object when it's not needed, so I'd have to somehow find or create one or be SOL. I could pass `None`, but what looks better and is easier to remember in your eyes?

`commands.timedelta_converter(my_string)`
`await commands.TimedeltaConverter().convert(None, my_string)`

mikeshardmind replied:
Actually, on that note @zephyrkul , I'd prefer if the actual conversion happened in a static method which was called from the convert to allow usages without the overhead of a creating a coroutine when being called manually. The only reason this needs to be a coroutine is it's use as a converter, so by offering manual usage as a feature, we should better support it as well.

Zephyrkul replied:
I'd prefer it if it wasn't an advanced `Converter` in the first place, which would remove the coro requirement (as well as the `ctx` requirement).
Did you mean to ping Laggron with that? Your phrasing seems to sound like you're addressing the person coding this PR.

mikeshardmind replied:
Actually, you're right. There's nothing used here which would need a coro or ctx, so This can easily just be a basic converter.

Redjumpman commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,80 @@
+
+    async def convert(self, ctx: "Context", argument: str) -> timedelta:
+        """
+        Convert manually a string to a :class:`datetime.timedelta` class.
```
Can we change this to` Manually convert a string ... ` instead of `Convert manually a string...`?

Redjumpman commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,80 @@
+
+        Raises
+        ------
+        ~discord.ext.commands.BadArgument
```
'~' character not needed.

laggron42 replied:
Isn't that character supposed to only show `BadArgument` in the docs instead of showing `discord.ext.commands.BadArgument`? 🤔

Redjumpman replied:
If that's the case, I'll resolve this then.

Redjumpman commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,80 @@
+if TYPE_CHECKING:
+    from .commands import Context
+
+__all__ = ["TimedeltaConverter"]
```
Not sure why dunder all is needed here. TimedeltaConverter is literally the only thing here. Typically you use all to "hide" some of the helper functions that build up the stuff you want developers to use.

laggron42 replied:
I followed `converters.py` as an example for this PR, and tried to match with their style, but if it's not needed I'll remove it 👍

Redjumpman commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,80 @@
+
+            @commands.command()
+            async def timer(self, ctx, time: commands.TimedeltaConverter):
+                await self.start_timer(time)
```
This is not a great example of how these deltas could be used. Perhaps change this very ambiguous classmethod to something like `await asyncio.sleep(time.seconds)`

Zephyrkul replied:
`time.total_seconds()` might be better so it picks up on days and up, which this converter converts.

Redjumpman replied:
True. `total_seconds` would be better.

Redjumpman commented:
I added some comments to this PR as well. I think this could be a helpful addition, but it needs some cleaning up.

laggron42 committed:
Remove __all__ to file

laggron42 committed:
Use "Manually convert" sentence

laggron42 committed:
Better example

laggron42 commented:
Travis is failing because of a Python 3.6.6 problem.

palmtree5 commented:
uhh, Travis isn't failing on this PR atm @retke

Tobotimus commented:
I restarted the job that failed, was just Travis having its periodic connection issues when downloading python

laggron42 committed:
Use a basic discord.py converter instead of an advanced one

laggron42 commented:
The 10 reviews are closed, is there something else to add?

laggron42 committed:
Remove unused imports

Zephyrkul commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,66 @@
+        .. code-block:: python3
+
+            @commands.command()
+            async def timer(self, ctx, time: commands.TimedeltaConverter):
```
This and line 42 need updating with the new syntax.

laggron42 committed:
Change the example's syntax

laggron42 committed:
Remove await from example

Zephyrkul commented on redbot/core/commands/converters.py:
```diff
@@ -0,0 +1,66 @@
+        .. code-block:: python3
+
+            async def convert_time(ctx: Context, text: str) -> datetime.timedelta:
+                time = await commands.timedelta_converter(text)
```
`timedelta_converter` is not awaitable.

laggron42 committed:
Merge syntax change

laggron42 committed:
Remove await from example

Tobotimus commented:
This was superseded by #2140
